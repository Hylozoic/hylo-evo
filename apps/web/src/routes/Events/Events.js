import cx from 'classnames'
import { get, isEmpty } from 'lodash/fp'
import React, { useEffect } from 'react'
import { Helmet } from 'react-helmet'
import { useTranslation } from 'react-i18next'
import { useDispatch, useSelector } from 'react-redux'
import { useNavigate, useParams, useLocation } from 'react-router-dom'
import Dropdown from 'components/Dropdown'
import GroupBanner from 'components/GroupBanner'
import Icon from 'components/Icon'
import Loading from 'components/Loading'
import NoPosts from 'components/NoPosts'
import PostCard from 'components/PostCard'
import ScrollListener from 'components/ScrollListener'
import fetchPosts from 'store/actions/fetchPosts'
import { FETCH_POSTS, FETCH_FOR_CURRENT_USER } from 'store/constants'
import presentPost from 'store/presenters/presentPost'
import getGroupForSlug from 'store/selectors/getGroupForSlug'
import getMe from 'store/selectors/getMe'
import getMyMemberships from 'store/selectors/getMyMemberships'
import { getHasMorePosts, getPosts } from 'store/selectors/getPosts'
import { createPostUrl } from 'util/navigation'
import { CENTER_COLUMN_ID } from 'util/scrolling'
import { updateTimeframe } from './Events.store'

import s from './Events.module.scss'
import viewStyles from '../../components/StreamViewControls/StreamViewControls.module.scss'

function Events (props) {
  const { t } = useTranslation()
  const dispatch = useDispatch()
  const navigate = useNavigate()
  const routeParams = useParams()
  const location = useLocation()

  const currentUser = useSelector(getMe)
  const currentUserHasMemberships = !isEmpty(useSelector(getMyMemberships))
  const groupSlug = routeParams.groupSlug
  const childPostInclusion = get('settings.streamChildPosts', currentUser) || 'yes'
  const group = useSelector(state => getGroupForSlug(state, groupSlug))
  const timeframe = useSelector(state => state.Events.timeframe)
  const context = props.context

  const fetchPostsParam = {
    childPostInclusion,
    afterTime: timeframe === 'future' ? new Date().toISOString() : null,
    beforeTime: timeframe === 'past' ? new Date().toISOString() : null,
    context,
    filter: 'event',
    order: timeframe === 'future' ? 'asc' : 'desc',
    slug: groupSlug,
    sortBy: 'start_time'
  }

  const posts = useSelector(state => getPosts(state, fetchPostsParam).map(p => presentPost(p, get('id', group))))
  const hasMore = useSelector(state => getHasMorePosts(state, fetchPostsParam))
  const membershipsPending = useSelector(state => state.pending[FETCH_FOR_CURRENT_USER])
  const pending = useSelector(state => state.pending[FETCH_POSTS])

  const fetchEvents = (offset) => dispatch(fetchPosts({ offset, ...fetchPostsParam }))
  const updateTimeframeAction = (newTimeframe) => dispatch(updateTimeframe(newTimeframe))
  const newPost = () => navigate(createPostUrl(routeParams, { newPostType: 'event' }))

  useEffect(() => {
    fetchEvents()
  }, [context, groupSlug, timeframe])

  const fetchMoreEvents = () => {
    if (pending || posts.length === 0 || !hasMore) return
    fetchEvents(posts.length)
  }

  const timeframeOptions = [
    { id: 'future', label: t('Upcoming Events') },
    { id: 'past', label: t('Past Events') }
  ]
  const upcomingText = t('upcoming')
  const pastText = t('past')

  if (!currentUser) return <Loading />
  if (membershipsPending) return <Loading />

  return (
    <div>
      <Helmet>
        <title>Events | {group ? `${group.name} |` : ''}Hylo</title>
      </Helmet>

      <GroupBanner
        group={group}
        currentUser={currentUser}
        type='event'
        context={context}
        newPost={newPost}
        routeParams={routeParams}
        querystringParams={location.search}
        currentUserHasMemberships={currentUserHasMemberships}
        icon='Events'
        label={t('Events')}
      />
      <div className={viewStyles.streamViewContainer}>
        <div className={viewStyles.streamViewCtrls}>
          <Dropdown
            className={viewStyles.dropdown}
            toggleChildren={
              <span className={viewStyles.dropdownLabel}>
                <Icon name='ArrowDown' />
                {timeframeOptions.find(o => o.id === timeframe).label}
              </span>
            }
            items={timeframeOptions.map(({ id, label }) => ({
              label,
              onClick: () => updateTimeframeAction(id)
            }))}
          />
        </div>
      </div>

      {pending && <Loading />}
      <div className={cx(s.eventsStream, { [s.collapsedState]: false })}>
        {!pending && posts.length === 0 ? <NoPosts message={t('No {{timeFrame}} events', { timeFrame: timeframe === 'future' ? upcomingText : pastText })} /> : ''}

        {posts.map(post => {
          const expanded = post.id === routeParams.postId
          const groupSlugs = post.groups.map(group => group.slug)
          return (
            <PostCard
              childPost={!groupSlugs.includes(groupSlug)}
              routeParams={routeParams}
              post={post}
              className={cx(s.eventCard, { [s.expanded]: expanded })}
              expanded={expanded}
              key={post.id}
            />
          )
        })}
      </div>
      <ScrollListener onBottom={fetchMoreEvents} elementId={CENTER_COLUMN_ID} />
    </div>
  )
}

export default Events
