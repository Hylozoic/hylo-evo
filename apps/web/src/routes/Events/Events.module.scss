.events-stream {
  margin: 0 auto;
  overflow: auto; // prevent margin collapse
}

.event-card {
  margin: 0 $space-4x $space-6x;
  box-shadow: 0 4px 10px rgb(35 65 91 / 20%);
}

.expanded {
  margin: 0 $space-2x $space-6x;
}

.create-group-prompt {
  text-align: center;
  font-size: 30px;
  align-items: center;
  justify-content: center;
  color: $color-rhino-50;
  padding-left: 100px;
  padding-right: 100px;
}

.sidebar-image {
  height: 150px;
  width: 100%;
  background: no-repeat center;
  background-size: contain;
}

.button {
  font-size: 14px;
  margin-bottom: 30px;
}
