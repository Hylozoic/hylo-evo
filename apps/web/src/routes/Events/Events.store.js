export const MODULE_NAME = 'Events'
export const UPDATE_TIMEFRAME = `${MODULE_NAME}/UPDATE_TIMEFRAME`

export function updateTimeframe (timeframe) {
  return {
    type: UPDATE_TIMEFRAME,
    payload: timeframe
  }
}

// reducer
const DEFAULT_STATE = {
  timeframe: 'future'
}

export default function (state = DEFAULT_STATE, action) {
  if (action.type === UPDATE_TIMEFRAME) {
    return {
      ...state,
      timeframe: action.payload
    }
  }
  return state
}
