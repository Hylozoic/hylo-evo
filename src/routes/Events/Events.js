import cx from 'classnames'
import { get } from 'lodash/fp'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { Helmet } from 'react-helmet'
import Dropdown from 'components/Dropdown'
import StreamBanner from 'components/StreamBanner'
import Icon from 'components/Icon'
import Loading from 'components/Loading'
import NoPosts from 'components/NoPosts'
import PostCard from 'components/PostCard'
import ScrollListener from 'components/ScrollListener'
import { CENTER_COLUMN_ID } from 'util/scrolling'
import s from './Events.scss'
import viewStyles from '../../components/StreamViewControls/StreamViewControls.scss'

const timeframeOptions = [
  { id: 'future', label: 'Upcoming Events' },
  { id: 'past', label: 'Past Events' }
]

export default class Events extends Component {
  static propTypes = {
    newPost: PropTypes.func,
    routeParams: PropTypes.object,
    querystringParams: PropTypes.object
  }

  static defaultProps = {
    routeParams: {},
    querystringParams: {}
  }

  componentDidMount () {
    this.props.fetchEvents()
  }

  componentDidUpdate (prevProps) {
    const { routeParams, fetchEvents } = this.props
    const { context, groupSlug } = routeParams

    if ((context && get('routeParams.context', prevProps) !== context) ||
        (groupSlug && get('routeParams.groupSlug', prevProps) !== groupSlug) ||
        prevProps.timeframe !== this.props.timeframe) {
      fetchEvents()
    }
  }

  fetchMoreEvents = () => {
    const { pending, posts, hasMore, fetchEvents } = this.props
    if (pending || posts.length === 0 || !hasMore) return
    fetchEvents(posts.length)
  }

  render () {
    const {
      collapsedState,
      currentUser,
      currentUserHasMemberships,
      group,
      groupSlug,
      membershipsPending,
      newPost,
      pending,
      posts,
      querystringParams,
      routeParams,
      timeframe, updateTimeframe
    } = this.props
    const { context } = routeParams

    if (!currentUser) return <Loading />
    if (membershipsPending) return <Loading />

    return (
      <div>
        <Helmet>
          <title>Events | {group ? `${group.name} |` : ''}Hylo</title>
        </Helmet>

        <StreamBanner
          group={group}
          currentUser={currentUser}
          type='event'
          context={context}
          newPost={newPost}
          routeParams={routeParams}
          querystringParams={querystringParams}
          currentUserHasMemberships={currentUserHasMemberships}
          icon='Events'
          label='Events'
        />
        <div styleName='viewStyles.stream-view-container'>
          <div styleName='viewStyles.stream-view-ctrls'>
            <Dropdown
              className={viewStyles.dropdown}
              toggleChildren={
                <span styleName='viewStyles.dropdown-label'>
                  <Icon name='ArrowDown' />
                  {timeframeOptions.find(o => o.id === timeframe).label}
                </span>
              }
              items={timeframeOptions.map(({ id, label }) => ({
                label,
                onClick: () => updateTimeframe(id)
              }))}
            />
          </div>
        </div>

        {pending && <Loading />}

        <div styleName={cx('s.events-stream', { [s.collapsedState]: collapsedState })}>
          {!pending && posts.length === 0 ? <NoPosts message={`No ${timeframe === 'future' ? 'upcoming' : 'past'} events`} /> : ''}

          {posts.map(post => {
            const expanded = post.id === routeParams.postId
            const groupSlugs = post.groups.map(group => group.slug)
            return (
              <PostCard
                childPost={!groupSlugs.includes(groupSlug)}
                routeParams={routeParams}
                post={post}
                styleName={cx('s.event-card', { 's.expanded': expanded })}
                expanded={expanded}
                key={post.id}
              />
            )
          })}
        </div>
        <ScrollListener onBottom={this.fetchMoreEvents} elementId={CENTER_COLUMN_ID} />

      </div>
    )
  }
}
