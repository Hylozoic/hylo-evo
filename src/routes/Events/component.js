import React from 'react'
import { Link } from 'react-router-dom'
import './component.scss'

const { string, bool } = React.PropTypes

export default function Events () {
  return <div styleName='event-list'>
    <div><Link to='/events/1'>Event 1</Link></div>
    <div><Link to='/events/2'>Event 2</Link></div>
    <div><Link to='/events/3'>Event 3</Link></div>
    <div><Link to='/events/4'>Event 4</Link></div>
  </div>
}
Events.propTypes = {
  label: string,
  color: string,
  active: bool,
  className: string
}
