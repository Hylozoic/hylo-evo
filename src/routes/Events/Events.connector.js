import { get, isEmpty } from 'lodash/fp'
import { push } from 'connected-react-router'
import { connect } from 'react-redux'
import fetchPosts from 'store/actions/fetchPosts'
import {
  FETCH_POSTS, FETCH_FOR_CURRENT_USER
} from 'store/constants'
import presentPost from 'store/presenters/presentPost'
import getGroupForCurrentRoute from 'store/selectors/getGroupForCurrentRoute'
import getRouteParam from 'store/selectors/getRouteParam'
import getMe from 'store/selectors/getMe'
import getMyMemberships from 'store/selectors/getMyMemberships'
import { getHasMorePosts, getPosts } from 'store/selectors/getPosts'
import { createPostUrl } from 'util/navigation'

import { updateTimeframe } from './Events.store'

export function mapStateToProps (state, props) {
  let group

  const routeParams = get('match.params', props)
  const location = get('location', props)
  const currentUser = getMe(state)
  const currentUserHasMemberships = !isEmpty(getMyMemberships(state))
  const groupSlug = getRouteParam('groupSlug', state, props)

  if (groupSlug) {
    group = getGroupForCurrentRoute(state, props)
  }

  const timeframe = state.Events.timeframe

  const fetchPostsParam = {
    afterTime: timeframe === 'future' ? new Date().toISOString() : null,
    beforeTime: timeframe === 'past' ? new Date().toISOString() : null,
    context: routeParams.context,
    filter: 'event',
    order: timeframe === 'future' ? 'asc' : 'desc',
    slug: groupSlug,
    sortBy: 'start_time'
  }

  // NOTE: In effort to better seperate the query caching from component details
  //       it's better (and necessary) in this case to send the fetch param then
  //       the raw props of the component.
  const posts = getPosts(state, fetchPostsParam).map(p => presentPost(p, get('id', group)))
  const hasMore = getHasMorePosts(state, fetchPostsParam)

  return {
    routeParams,
    location,
    currentUser,
    currentUserHasMemberships,
    fetchPostsParam,
    groupSlug,
    group,
    hasMore,
    membershipsPending: state.pending[FETCH_FOR_CURRENT_USER],
    posts,
    pending: state.pending[FETCH_POSTS],
    timeframe
  }
}

export function mapDispatchToProps (dispatch, props) {
  const routeParams = get('match.params', props)

  return {
    fetchEvents: (params) => (offset) => dispatch(fetchPosts({ offset, ...params })),
    updateTimeframe: timeframe => dispatch(updateTimeframe(timeframe)),
    newPost: () => dispatch(push(createPostUrl(routeParams, { newPostType: 'event' })))
  }
}

export function mergeProps (stateProps, dispatchProps, ownProps) {
  const { fetchPostsParam } = stateProps

  return {
    ...ownProps,
    ...stateProps,
    ...dispatchProps,
    fetchEvents: dispatchProps.fetchEvents(fetchPostsParam)
  }
}

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)
