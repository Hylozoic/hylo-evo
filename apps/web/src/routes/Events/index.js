import component from './Events'

export default component
