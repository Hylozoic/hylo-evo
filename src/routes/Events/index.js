import CSSModules from 'react-css-modules'
import component from './component'
import styles from './component.scss'

export default CSSModules(styles, {allowMultiple: true})(component)
