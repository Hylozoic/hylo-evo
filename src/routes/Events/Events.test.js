import React from 'react'
import { shallow } from 'enzyme'
import Events from './Events'

describe('Events', () => {
  it('renders a post list', () => {
    const posts = [{ id: 1, groups: [] }, { id: 2, groups: [] }, { id: 3, groups: [] }]
    const wrapper = shallow(<Events
      currentUser={{ id: 1 }}
      posts={posts}
      fetchEvents={jest.fn()}
      timeframe='future'
      currentUserHasMemberships
    />)
    expect(wrapper.find('Connect(PostCard)').length).toEqual(3)
    expect(wrapper).toMatchSnapshot()
  })

  it('displays the regular StreamBanner', () => {
    const props = {
      currentUser: { id: 1 },
      posts: [],
      timeframe: 'future',
      fetchEvents: jest.fn()
    }
    const wrapper = shallow(<Events {...props} />)
    expect(wrapper.find('StreamBanner')).toHaveLength(1)
  })
})
