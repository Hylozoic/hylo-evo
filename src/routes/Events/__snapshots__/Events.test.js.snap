// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events renders a post list 1`] = `
<div>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Events | 
      Hylo
    </title>
  </HelmetWrapper>
  <StreamBanner
    currentUser={
      Object {
        "id": 1,
      }
    }
    currentUserHasMemberships={true}
    icon="Events"
    label="Events"
    querystringParams={Object {}}
    routeParams={Object {}}
    type="event"
  />
  <div
    data-stylename="viewStyles.stream-view-container"
  >
    <div
      data-stylename="viewStyles.stream-view-ctrls"
    >
      <Dropdown
        items={
          Array [
            Object {
              "label": "Upcoming Events",
              "onClick": [Function],
            },
            Object {
              "label": "Past Events",
              "onClick": [Function],
            },
          ]
        }
        toggleChildren={
          <span
            data-stylename="viewStyles.dropdown-label"
          >
            <Icon
              name="ArrowDown"
            />
            Upcoming Events
          </span>
        }
      />
    </div>
  </div>
  <div
    data-stylename="s.events-stream"
  >
    <Connect(PostCard)
      childPost={true}
      data-stylename="s.event-card"
      expanded={false}
      key="1"
      post={
        Object {
          "groups": Array [],
          "id": 1,
        }
      }
      routeParams={Object {}}
    />
    <Connect(PostCard)
      childPost={true}
      data-stylename="s.event-card"
      expanded={false}
      key="2"
      post={
        Object {
          "groups": Array [],
          "id": 2,
        }
      }
      routeParams={Object {}}
    />
    <Connect(PostCard)
      childPost={true}
      data-stylename="s.event-card"
      expanded={false}
      key="3"
      post={
        Object {
          "groups": Array [],
          "id": 3,
        }
      }
      routeParams={Object {}}
    />
  </div>
  <ScrollListener
    elementId="center-column"
    onBottom={[Function]}
  />
</div>
`;
